### ISSUE TITLE
Research: current-main Needle 2.0.12 compatibility canaries

### ISSUE THREAD

**Issue by TeaShaman-cyber** [TeaShaman-cyber/theseus-needle-lab#44]
## Research question

Which findings from the pinned `cactus-needle==2.0.8` Theseus experiments still reproduce on current upstream Needle, and which were artifacts of pre-2.0.12 training/deployment behavior?

This issue is a compatibility/reproduction track. It must not rewrite the historical Stage B result in #26 or delay the preregistered Stage C experiment.

## Current upstream authority

At the start of this track:

- upstream repository: `cactus-compute/needle`;
- current inspected commit: `cd2f97642643e70832d8c6f8d9ba998b9c168de9`;
- package version: `2.0.12`;
- upstream issue #91 reports that 2.0.12 fixed both the fine-tuning/native-inference template mismatch and deployment-numerics mismatch via QAT.

## FACT

1. Stage B #26 used `cactus-needle==2.0.8` and evaluated the built tuned `.cact`, not a separate float-LoRA control.
2. Therefore Stage B remains valid as an end-to-end 2.0.8 deployed-artifact result, but its applicability collapse cannot be attributed uniquely to the SFT objective.
3. Current-main `needle finetune` has no user-facing seed in the inspected upstream commit, while the implementation combines:
   - LoRA init with fixed PRNG seed `0`;
   - validation split with fixed NumPy seed `0`;
   - epoch shuffling through uncontrolled global `np.random.permutation`.
4. Two clean-process epoch-shuffle canaries produced different permutations.
5. Current-main `render_example()` serializes tool-schema dicts with insertion order preserved. Two semantically equivalent JSON tool objects with different key insertion order produced different exact prompt SHA256 values:
   - `c56bf4444d4938a1467e6e159b9b741fa1d1c739a74feeb7b7b267d895ffe55f`
   - `599443d94b8739497d69b6e090b1973cd236914c77797e946466af4a92d7ef51`

The prompt-byte difference alone is not evidence of a behavioral regression.

## UNKNOWN

- Whether the Stage B applicability collapse reproduces after retraining on 2.0.12/QAT.
- Whether float-adapter and built `.cact` behavior remain materially different on the Stage B corpus.
- Whether schema insertion-order differences cause measurable output changes on current-main weights.
- Whether canonicalizing schemas would improve reproducibility without harming compatibility with the model's learned serialization prior.

## Work streams

```mermaid
flowchart TD
    A[Historical Stage B: Needle 2.0.8] --> B[Upstream #91 causal correction]
    B --> C[Current-main 2.0.12 canaries]
    C --> D[Seed / training-order reproducibility]
    C --> E[Float LoRA vs built .cact]
    C --> F[Schema insertion-order behavior]
    D --> G[Small upstream PR if verified]
    E --> H[Update Stage B causal interpretation]
    F --> I[Docs/test/PR only if behavioral effect reproduces]
```

### A. Finetune seed surface

Add a backward-compatible `needle finetune --seed`, default `0`, and use it consistently for LoRA initialization, validation selection, and epoch shuffling. Persist the realized seed in adapter metadata. Upstream PR only after RED/GREEN and current test-suite verification.

### B. Float-vs-built canary

Retrain a bounded Stage B-compatible slice on 2.0.12 and evaluate at minimum:

```text
base
float LoRA / JAX path
built tuned .cact / native path
```

Hold dataset, seed, schema bytes, and evaluation cases fixed. This canary is required before attributing any current-main applicability effect to training versus export/deployment numerics.

### C. Schema-order canary

Use semantically identical schemas with different insertion order. First prove exact rendered-byte difference (already observed), then test whether current-main behavior changes under paired same-run evaluation. Do not open an upstream bug from byte differences alone.

## Guardrails

- Historical Stage B evidence remains immutable.
- No claim that current upstream `main` reproduces 2.0.8 behavior until measured.
- No uncontrolled training sweeps.
- Heldout remains evaluation-only.
- Separate operational reproducibility findings from scientific model-quality findings.
- Prefer one small upstream contribution at a time rather than bundling unrelated changes.

## Related public upstream activity

- `cactus-compute/needle#61`: downstream evaluation discussion; Theseus Stage B result shared with corrected 2.0.8 causal boundary.
- `cactus-compute/needle#91`: template/QAT/deployment-numerics correction relevant to Stage B interpretation.
- local public downstream note: PR #43.

---

— Semyon Poklad & **Шут (Jester)**, Theseus research project

**Comment by TeaShaman-cyber** [comment]
## Upstream acceptance checkpoint — seed/reproducibility work landed

The bounded upstream contribution from Work stream A has been accepted and merged by the Needle maintainers:

- upstream PR: cactus-compute/needle#108 — `finetune: make training seed explicit`
- merged: 2026-09-07T18:12:17Z
- upstream merge commit: `cd2eae34b96e20b15b1c0fcb7e9fe683ed6a675b`
- contributed head: `21b414f438a9c1f653cb2e509f92db3474b81d92`

The merged change exposes `needle finetune --seed`, applies the seed consistently to LoRA initialization, validation selection, and epoch shuffling, and records the realized seed in adapter metadata.

Independent current-main readback also shows that the next upstream main commit (`53df049c4a1a82fca1027b81f9ff21336dfb0861`, #109) has the #108 merge commit as its parent. So this is not merely a closed PR: the change is in the current upstream lineage.

### Research-state update

The original FACT in this issue — that the inspected pre-merge upstream `main` had no user-facing seed and used uncontrolled global NumPy shuffling — remains valid for the pinned historical commit `cd2f97642643e70832d8c6f8d9ba998b9c168de9`, but is now **superseded for current upstream main** by #108.

```text
seed surface missing on inspected old main = HISTORICAL FACT
upstream reproducibility fix accepted      = VERIFIED
upstream current lineage contains fix       = VERIFIED
Work stream A upstream contribution         = COMPLETE
```

This does not answer the remaining scientific questions in #44: float-LoRA vs built `.cact`, Stage B applicability on current Needle, and schema-order behavioral sensitivity remain separate canaries.

— Семён & Шут

**Comment by TeaShaman-cyber** [comment]
## Currentness checkpoint — release identity and process invariants (2026-09-10)

A bounded currentness audit was run before any new training or runtime canary. No dependencies were installed and no model execution was performed.

### Observed current upstream state

- PyPI currently publishes `cactus-needle==2.0.13`.
- Git tag `v2.0.13` points to `40975f9440ed995a91e9e372f6ab47a478b84108`.
- That release commit has current upstream `main` SHA `53df049c4a1a82fca1027b81f9ff21336dfb0861` as its direct parent.
- The release commit changes only version metadata (`2.0.12 -> 2.0.13`) in `needle/__init__.py` and `pyproject.toml`; the functional code is the same as that inspected `main` parent.
- There is no Git tag `v2.0.12` even though PyPI contains package version 2.0.12. Therefore a display/package version alone is not a sufficient source identity.
- Published PyPI wheel `cactus_needle-2.0.13-py3-none-any.whl` SHA256 observed in this audit: `5f5975aa7451ac8db9169fb75dbcf24a7a3dc4118ae82b6ffd405836fa91a823`.
- Key files in that wheel (`needle/__init__.py`, `needle/model/finetune.py`, `needle/cli.py`) byte-match the `v2.0.13` Git tag.

### #91 correction is present in the released code

Static inspection of `v2.0.13` verifies:

- `render_example()` emits `<think>\n...\n</think>\n` for non-empty reasoning;
- QAT defaults to `auto` and records/uses the checkpoint quantization scheme;
- `needle finetune --seed` exists, uses a single seeded RNG for epoch permutations, seeds LoRA initialization, and records the realized seed.

This supports the upstream #91 correction as present in the released implementation, but does **not** prove downstream Stage B behavior is fixed. Upstream explicitly says old adapters must be retrained.

### Multi-generation boundary

`v2.0.13` contains Needle 2 / Needle 3 dispatch code, but it does not silently move the old default inference path to Needle 3:

- no tuned `weights=` -> generation 2;
- tuned `.cact` -> generation selected from the archive format tag;
- a v3 archive is not routed through the v2 engine.

Therefore a bounded Needle-2 reproduction remains possible if the exact release/artifact identities are pinned.

### Process invariants exposed by this audit

1. **Depend on a finding, not completion of an issue.** A downstream experiment that needs only the current inference baseline must not wait for every branch of #44.
2. **Version label != immutable runtime identity.** Bind at least package/release version + exact Git/source identity + executable/model artifact digest where applicable.
3. **Source tag != published artifact until bridged by evidence.** The wheel/tag match above is such a bridge for the inspected files, not a universal guarantee for every distribution surface.
4. **Historical evidence is immutable; currentness narrows interpretation.** Stage B remains a valid 2.0.8 deployed-artifact result. New upstream fixes change how broadly its cause may be generalized; they do not rewrite the old result.
5. **A changed upstream capability matters only when it lies on the causal path of the experiment.** Training-only seed changes are not automatically prerequisites for inference-only experiments.
6. **Stable release and moving `main` are distinct experimental targets.** Do not call one “current” without naming which authority is intended.
7. **Open issue != current design.** Re-check the smallest authoritative upstream state immediately before execution rather than trusting an old title/body snapshot.

### Current disposition

For purposes that only need a stable current Needle-2 inference baseline, `cactus-needle==2.0.13` / tag `40975f9440ed995a91e9e372f6ab47a478b84108` is now a concrete candidate baseline, with the published wheel digest above.

This checkpoint does not complete #44. The remaining scientific questions — especially retrained 2.0.13 Stage B behavior, float-vs-built behavior, and schema-order behavioral sensitivity — remain separate canaries and should be run only when they are needed for their own claims.

`CURRENTNESS CHECKPOINT VERIFIED / NO TRAINING AUTHORIZED / NO MODEL-QUALITY CLAIM`

**Comment by TeaShaman-cyber** [comment]
## Downstream dependency split — baseline finding is now consumable

The 2026-09-10 currentness checkpoint is sufficient for downstream work that only needs a stable current Needle-2 inference baseline.

For that narrow dependency, the finding is:

```text
stable Needle-2 candidate baseline
= cactus-needle==2.0.13
= Git tag 40975f9440ed995a91e9e372f6ab47a478b84108
= published wheel sha256 5f5975aa7451ac8db9169fb75dbcf24a7a3dc4118ae82b6ffd405836fa91a823
```

This does **not** mean #44 is complete. The remaining questions are independent claims:

- whether retrained Stage B behavior changes on 2.0.13;
- float-LoRA vs built `.cact` behavior;
- schema-order behavioral sensitivity.

They should be required only by downstream claims that causally depend on them.

Process rule exposed by this split:

`dependency on finding/checkpoint != dependency on completion of the whole issue`

For #46 specifically, the Needle-baseline prerequisite is now satisfied. Any remaining HOLD there must come from its own context-factor design, not from unfinished branches of #44.

`DOWNSTREAM BASELINE FINDING = SATISFIED / #44 REMAINS OPEN FOR INDEPENDENT CANARIES`

**Comment by TeaShaman-cyber** [comment]
## Currentness checkpoint — 2026-09-22

This issue remains valid historical evidence for the Needle 2.0.12/current-main transition that it actually measured and designed around.

Current operational disposition:

`STALE_FOR_CURRENTNESS / SUPERSEDED_FOR_CURRENTNESS_BY #61 / HISTORICAL_RECEIPTS_PRESERVED`

Needle 3 is now the live upstream line. PyPI currently reports `cactus-needle==3.0.4` (released 2026-09-21), and upstream `cactus-compute/needle#134` reports a JAX merged-LoRA versus built-`.cact` behavioral divergence on 3.0.2. That upstream report is a reason to reprobe, not proof about 3.0.4 or later.

#61 is therefore the current reprobe authority for the deployment-equivalence question. #44 is not closed or rewritten; its 2.x evidence remains a predecessor receipt.
